From 502fcf1198f15b18bea184eab0397c8f90223ccf Mon Sep 17 00:00:00 2001
From: Yuval Caduri <cyuval@marvell.com>
Date: Wed, 7 Feb 2018 10:48:53 +0200
Subject: [PATCH 2233/2241] fix: net: mvpp2x: open cls once for userspace port

Fixes rss in userspace.
For userspace interfaces mv_pp2x_open_cls() must not be called
during interface up/down because it overrides userspace cls config.
But it must be called once, before userspace starts,for rss to work
in userspace.

Change-Id: I4833e7b07bb1bff1310e5f9b6aebb531eed86092
Signed-off-by: Yuval Caduri <cyuval@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/50166
Reviewed-by: Stefan Chulski <stefanc@marvell.com>
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
---
 drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
index dbc7cbc..6bac51b 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
@@ -4264,10 +4264,6 @@ int mv_pp2x_open_cls(struct net_device *dev)
 		return err;
 	}
 
-	/* For musdk ports do not updated classifier/rss, only parser. */
-	if (port->flags & MVPP2_F_IF_MUSDK)
-		return 0;
-
 	err = mv_pp2x_update_flow_info(hw);
 	if (err) {
 		netdev_err(port->dev, "cannot update flow info\n");
@@ -4450,9 +4446,11 @@ int mv_pp2x_open(struct net_device *dev)
 	/* Before rxq and port init, all ingress packets should be blocked
 	 *  in classifier
 	 */
-	err = mv_pp2x_open_cls(dev);
-	if (err)
-		goto err_free_all;
+	if (!(port->flags & MVPP2_F_IF_MUSDK)) {
+		err = mv_pp2x_open_cls(dev);
+		if (err)
+			goto err_free_all;
+	}
 
 	return 0;
 
@@ -5573,6 +5571,9 @@ int mv_pp2x_port_musdk_set(void *netdev_priv)
 	port->num_rx_queues = 0;
 	port->num_tx_queues = 0;
 
+	/* For MUSDK port, run cls_open once */
+	mv_pp2x_open_cls(port->dev);
+
 	port->flags |= MVPP2_F_IF_MUSDK;
 	if (was_running) {
 		dev_open(port->dev);
-- 
2.7.4

